File: chenyijie/08/src/codewriter.py
#写入代码
from constant import *

class CodeWriter(object):
    """docstring fo CodeWriter."""
    def __init__(self, filename):

        self._filename = filename
        self._eq_label_num = 1
        self._ne_label_num = 1
        self._label_num = 1


    def setfilename(self):
        self._file = open(self._filename, 'w')


    def writeCommands(self,commands):
        for n in commands:
            self._file.write(n+'\n')
            print(n)


    def eq_label(self):
        return 'LABEL_SATISFY_'+str(self._eq_label_num)


    def ne_label(self):
        return 'LABEL_DISSATISIFY_'+str(self._ne_label_num)


    def label(self):
        return 'LABEL_'+str(self._label_num)


    def writeArithmetic(self, command):
        if command == 'add':
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=D+A','@SP','A=M','M=D','@SP','M=M+1'])
            #                   取出栈顶元素                      索引到第二个元素                相加              放回栈顶             索引归位
        if command == 'sub':
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A-D','@SP','A=M','M=D','@SP','M=M+1'])
            #                   取出栈顶元素                      索引到第二个元素                相减              放回栈顶             索引归位
        if command == 'eq':#x=y
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A-D',
            '@'+self.eq_label(),'D;JEQ','@SP','A=M','M=0',#如果相等跳转，否则设置为0
            '@'+self.ne_label(),'0;JMP',#跳至不相等的情况
            '('+self.eq_label()+')','@SP','A=M','M=-1',#相等，置为-1
            '('+self.ne_label()+')','@SP','M=M+1'])
            #让标签数增加，防止重复出现
            self._eq_label_num = self._eq_label_num + 1
            self._ne_label_num = self._ne_label_num + 1

        if command == 'neg':#-y
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','D=-D','M=D','@SP','M=M+1'])
        if command == 'not':
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','D=!D','M=D','@SP','M=M+1'])

        if command == 'gt':#x>y
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A-D',
            '@'+self.eq_label(),'D;JGT','@SP','A=M','M=0',
            '@'+self.ne_label(),'0;JMP',
            '('+self.eq_label()+')','@SP','A=M','M=-1',
            '('+self.ne_label()+')','@SP','M=M+1'])
            self._eq_label_num = self._eq_label_num + 1
            self._ne_label_num = self._ne_label_num + 1
        if command == 'lt':#x<y
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A-D',
            '@'+self.eq_label(),'D;JLT','@SP','A=M','M=0',
            '@'+self.ne_label(),'0;JMP',
            '('+self.eq_label()+')','@SP','A=M','M=-1',
            '('+self.ne_label()+')','@SP','M=M+1'])
            self._eq_label_num = self._eq_label_num + 1
            self._ne_label_num = self._ne_label_num + 1
        if command == 'and':
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A&D','@SP','A=M','M=D','@SP','M=M+1'])
        if command == 'or':
            self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@SP','M=M-1','@SP','A=M','A=M','D=A|D','@SP','A=M','M=D','@SP','M=M+1'])


    def writePushPoP(self, commandtype, arg1, arg2):
        if commandtype == C_PUSH:
            if arg1 == 'constant':
                self.writeCommands(['@'+arg2, 'D=A','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'local':
                self.writeCommands(['@'+arg2, 'D=A','@LCL','A=M','AD=D+A','D=M','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'that':
                self.writeCommands(['@'+arg2, 'D=A','@THAT','A=M','AD=D+A','D=M','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'this':
                self.writeCommands(['@'+arg2, 'D=A','@THIS','A=M','AD=D+A','D=M','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'argument':
                self.writeCommands(['@'+arg2, 'D=A','@ARG','A=M','AD=D+A','D=M','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'temp':
                self.writeCommands(['@'+arg2, 'D=A','@TEMP','A=M','AD=D+A','D=M','@SP','A=M','M=D','@SP','M=M+1'])
            if arg1 == 'static':
                self.writeCommands(['@'+arg2, 'D=M','@SP','A=M','M=D','@SP','M=M+1'])
        if commandtype == C_POP:
            if arg1 == 'local':
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@LCL','A=M','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'argument':
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@ARG','A=M','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'this':
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@THIS','A=M','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'that':
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@THAT','A=M','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'temp':
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@TEMP','A=M','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'pointer':#为什么会不一样？
                self.writeCommands(['@SP','M=M-1','@'+arg2, 'D=A','@3','AD=D+A','@R13','M=D','@SP','A=M','D=M','@R13','A=M','M=D'])
            if arg1 == 'static':#为什么会不一样？
                self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@'+arg2,'M=D'])
    def writeInit(self):
        #SP=256
        self.writeCommands(['@256','D=A','@SP','M=D'])
        #call Sys.init
        self.writeCall('Sys.init', '0')


    def writeLabel(self, label):
        self.writeCommands(['('+label+')'])


    def writeGoto(self, label):
        self.writeCommands(['@'+label,'0;JMP'])


    def writeIf(self, label):#将布尔表达式的值从栈顶弹出，如果非0则跳转
        self.writeCommands(['@SP','M=M-1','@SP','A=M','D=M','@'+label,'D;JNE'])


    def writeCall(self, FunctionName, NumArgs):
        NumArgs = int(NumArgs)
        #push return-adress
        self.writeCommands(['@'+self.label(),'D=A','@SP','A=M','M=D','@SP','M=M+1'])
        #push LCL
        self.writeCommands(['@LCL','D=M','@SP','A=M','M=D','@SP','M=M+1'])
        #push ARG
        self.writeCommands(['@ARG','D=M','@SP','A=M','M=D','@SP','M=M+1'])
        #push THIS
        self.writeCommands(['@THIS','D=M','@SP','A=M','M=D','@SP','M=M+1'])
        #push THAT
        self.writeCommands(['@THAT','D=M','@SP','A=M','M=D','@SP','M=M+1'])
        #ARG=SP-n-5
        self.writeCommands(['@'+str(5+NumArgs),'D=A','@SP','A=M','AD=A-D','@ARG','M=D'])
        #LCL=SP
        self.writeCommands(['@SP','D=M','@LCL','M=D'])
        #goto f
        self.writeCommands(['@'+FunctionName,'0;JMP'])
        #(return-address)
        self.writeCommands(['('+self.label()+')'])

        self._label_num = self._label_num+1


    def writeReturn(self):
        #FRAME=LCL
        self.writeCommands(['@LCL','D=M','@FRAME','M=D'])
        #RET=*(FRAME-5)
        self.writeCommands(['@5','A=D-A','D=M','@RET','M=D'])
        #*ARG=POP()
        self.writeCommands(['@SP','M=M-1','@ARG','AD=M','@R15','M=D','@SP','A=M','D=M','@R15','A=M','D=M'])
        #SP=ARG+1
        self.writeCommands(['@ARG','D=M','@SP','M=D+1'])
        #THAT=*(FRAME-1)
        self.writeCommands(['@FRAME','D=M','D=D-1','@FRAME','M=D','A=D','D=M','@THAT','M=D'])
        #THIS=*(FRAME-2)
        self.writeCommands(['@FRAME','D=M','D=D-1','@FRAME','M=D','A=D','D=M','@THIS','M=D'])
        #ARG=*(FRAME-3)
        self.writeCommands(['@FRAME','D=M','D=D-1','@FRAME','M=D','A=D','D=M','@ARG','M=D'])
        #LCL=*(FRAME-4)
        self.writeCommands(['@FRAME','D=M','D=D-1','@FRAME','M=D','A=D','D=M','@LCL','M=D'])
        #goto RET
        self.writeCommands(['@RET','A=M','0;JMP'])


    def witreFunction(self, FunctionName, NumLocals):
        self.writeLabel(FunctionName)#为函数入口声明一个标签
        NumLocals = int(NumLocals)
        i = 0
        while i<NumLocals:#重复k次，初始化k个局部变量为0
            i=i+1
            self.writeCommands(['@0','D=A','@SP','A=M','M=D','@SP','M=M+1'])

File: WangYida/projects/08/VMtranslator/VMtranslator/Codewriter.cpp
#include<iostream>
#include<string>
#include<fstream>
#include"VMtranslator.h"
#include "Parser.h"
#include "Codewriter.h"
#include "Symboltable.h"

using namespace std;

void Init(){
	std::ofstream fout(OutputName, std::ios::app);

	fout << "@" << spinit <<std::endl;
	Out("D=A");
	Out("@SP");
	Out("M=D");
	fout << "@Sys.init" << std::endl << "0;JMP" << std::endl; 
}

void Out(char *s){
	std::ofstream fout(OutputName, std::ios::app);
	fout << s;
	fout << std::endl;
}

void Deal(){
	//cout << OutputName << std::endl << filename;
	if (CommandType() >= 101 && CommandType() <= 109)
		DealWithArithmetic();
	if(CommandType() == 110 || CommandType() == 111)
		DealWithMemory();
	if (CommandType() >= 112 && CommandType() <= 114)
		DealWithProgramFlow();
	if (CommandType() == 115)
		DealWithFunction();
	if (CommandType() == 116)
		DealWithCallFunction();
	if (CommandType() == 117)
		DealWithReturn();
	//to do

}

void DealWithArithmetic(){
	switch(CommandType() ){
	case 101 : {		//ADD_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D+A");
		Out("@SP");
		Out("A=M");
		Out("M=D");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	case 102 : {		//SUB_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D-A");
		Out("D=-D");
		Out("@SP");
		Out("A=M");
		Out("M=D");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	case 103 : {			//NEG_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("M=-M");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	case 104 : {			//AND
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D&A");
		Out("@SP");
		Out("A=M");
		Out("M=D");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	case 105 : {			//NOT_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("M=!M");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	case 106 : {			//EQ_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D-A");
		AtNormalSymbol(At_NormalSymbol);
		Out("D;JEQ");
		Out("@SP");
		Out("A=M");
		Out("M=0");
		Out("@SP");
		Out("M=M+1");
		AtNormalSymbol(At_NormalSymbol);
		Out("0;JMP");
		AddNormalSymbol(Add_NormalSymbol);
		Out("@SP");
		Out("A=M");
		Out("M=-1");
		Out("@SP");
		Out("M=M+1");
		AddNormalSymbol(Add_NormalSymbol);
		break;
			   }
	case 107 : {			//GT_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D-A");
		Out("D=-D");
		AtNormalSymbol(At_NormalSymbol);
		Out("D;JGT");
		Out("@SP");
		Out("A=M");
		Out("M=0");
		Out("@SP");
		Out("M=M+1");
		AtNormalSymbol(At_NormalSymbol);
		Out("0;JMP");
		AddNormalSymbol(Add_NormalSymbol);
		Out("@SP");
		Out("A=M");
		Out("M=-1");
		Out("@SP");
		Out("M=M+1");
	    AddNormalSymbol(Add_NormalSymbol);
		break;
			   }
	case 108 : {			//LT_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D-A");
		Out("D=-D");
		AtNormalSymbol(At_NormalSymbol);
		Out("D;JLT");
		Out("@SP");
		Out("A=M");
		Out("M=0");
		Out("@SP");
		Out("M=M+1");
		AtNormalSymbol(At_NormalSymbol);
		Out("0;JMP");
		AddNormalSymbol(Add_NormalSymbol);
		Out("@SP");
		Out("A=M");
		Out("M=-1");
		Out("@SP");
		Out("M=M+1");
		AddNormalSymbol(Add_NormalSymbol);
		break;
			   }
	case 109 : {			//OR_
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("D=M");
		Out("@SP");
		Out("M=M-1");
		Out("@SP");
		Out("A=M");
		Out("A=M");
		Out("D=D|A");
		Out("@SP");
		Out("A=M");
		Out("M=D");
		Out("@SP");
		Out("M=M+1");
		break;
			   }
	default :
		break;
	}
}

void DealWithMemory(){
	if(CommandType() == 110){		//PUSH_
		
		if(strcmp(secondword, "constant") == 0){				//constant
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		//if(instruction[4] == 'a' && instruction[5] == 'r' && instruction[6] == 'g'){	//argument
		if(strcmp(secondword, "argument") == 0){
			//for(int i = 0; i < (strlen(instruction) - 12); i++)
			//	temp[i] = instruction[i + 12];
			//temp[strlen(instruction) - 12] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@ARG");
			Out("A=M");
			Out("A=D+A");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}
		
		//if(instruction[4] == 't' && instruction[5] == 'h' && instruction[6] == 'i'){	//this
		if(strcmp(secondword, "this") == 0){
			//for(int i = 0; i < (strlen(instruction) - 8); i++)
			//	temp[i] = instruction[i + 8];
			//temp[strlen(instruction) - 8] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@THIS");
			Out("A=M");
			Out("A=D+A");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		//if(instruction[4] == 't' && instruction[5] == 'h' && instruction[6] == 'a'){	//that
		if(strcmp(secondword, "that") == 0){
			//for(int i = 0; i < (strlen(instruction) - 8); i++)
		//		temp[i] = instruction[i + 8];
			//temp[strlen(instruction) - 8] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@THAT");
			Out("A=M");
			Out("A=D+A");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		//if(instruction[4] == 'l' && instruction[5] == 'o' && instruction[6] == 'c'){	//local
		if(strcmp(secondword, "local") == 0){
			//for(int i = 0; i < (strlen(instruction) - 9); i++)
			//	temp[i] = instruction[i + 9];
			//temp[strlen(instruction) - 9] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@LCL");
			Out("A=M");
			Out("A=D+A");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		
		//if(instruction[4] == 't' && instruction[5] == 'e' && instruction[6] == 'm'){	//temp
		if(strcmp(secondword, "temp") == 0){
			//for(int i = 0; i < (strlen(instruction) - 8); i++)
			//	temp[i] = instruction[i + 8];
			//temp[strlen(instruction) - 8] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@5");
			Out("A=D+A");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		//if(instruction[4] == 's' && instruction[5] == 't' && instruction[6] == 'a'){	//static
		if(strcmp(secondword, "static") == 0){
			//for(int i = 0; i < (strlen(instruction) - 10); i++)
			//	temp[i] = instruction[i + 10];
			//temp[strlen(instruction) - 10] = '\0';
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << filename << "." << thirdword << std::endl;
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}

		//if(instruction[4] == 'p' && instruction[5] == 'o' && instruction[6] == 'i'){	//pointer
		if(strcmp(secondword, "pointer") == 0){
			//for(int i = 0; i < (strlen(instruction) - 11); i++)
			//	temp[i] = instruction[i + 11];
			//temp[strlen(instruction) - 11] = '\0';

			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword <<std::endl;
			Out("D=A");
			Out("@3");
			Out("D=D+A");
			Out("A=D");
			Out("D=M");
			Out("@SP");
			Out("A=M");
			Out("M=D");
			Out("@SP");
			Out("M=M+1");
		}


	}
	if(CommandType() == 111){		//POP_

		//if(instruction[3] == 'a' && instruction[4] == 'r' && instruction[5] == 'g'){	//argument
			//for(int i = 0; i < (strlen(instruction) - 11); i++)
			//	temp[i] = instruction[i + 11];
			//temp[strlen(instruction) - 11] = '\0';
		if(strcmp(secondword, "argument") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@ARG");
			Out("A=M");
			Out("D=D+A");
			Out("@R14");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("M=D");
			Out("@R14");
			Out("A=M");
			Out("M=D");
		}

		//if(instruction[3] == 't' && instruction[4] == 'h' && instruction[5] == 'i'){	//this
			//for(int i = 0; i < (strlen(instruction) - 7); i++)
			//	temp[i] = instruction[i + 7];
			//temp[strlen(instruction) - 7] = '\0';
		if(strcmp(secondword, "this") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@THIS");
			Out("A=M");
			Out("D=D+A");
			Out("@R14");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("M=D");
			Out("@R14");
			Out("A=M");
			Out("M=D");
		}

		//if(instruction[3] == 't' && instruction[4] == 'h' && instruction[5] == 'a'){	//that
		//	for(int i = 0; i < (strlen(instruction) - 7); i++)
		//		temp[i] = instruction[i + 7];
		//	temp[strlen(instruction) - 7] = '\0';
		if(strcmp(secondword, "that") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@THAT");
			Out("A=M");
			Out("D=D+A");
			Out("@R14");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("M=D");
			Out("@R14");
			Out("A=M");
			Out("M=D");
		}

		//if(instruction[3] == 'l' && instruction[4] == 'o' && instruction[5] == 'c'){	//local
		//	for(int i = 0; i < (strlen(instruction) - 8); i++)
		//		temp[i] = instruction[i + 8];
		//	temp[strlen(instruction) - 8] = '\0';
		if(strcmp(secondword, "local") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@LCL");
			Out("A=M");
			Out("D=D+A");
			Out("@R14");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("M=D");
			Out("@R14");
			Out("A=M");
			Out("M=D");
		}

		
		//if(instruction[3] == 't' && instruction[4] == 'e' && instruction[5] == 'm'){	//temp
		//	for(int i = 0; i < (strlen(instruction) - 7); i++)
		//		temp[i] = instruction[i + 7];
		//	temp[strlen(instruction) - 7] = '\0';
		if(strcmp(secondword, "temp") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword << std::endl;
			Out("D=A");
			Out("@5");
			Out("D=D+A");
			Out("@R13");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("A=M");
			Out("M=D");
		}

	//	if(instruction[3] == 's' && instruction[4] == 't' && instruction[5] == 'a'){	//static
		//	for(int i = 0; i < (strlen(instruction) - 9); i++)
		//		temp[i] = instruction[i + 9];
		//	temp[strlen(instruction) - 9] = '\0';
		if(strcmp(secondword, "static") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << filename << "." << thirdword << std::endl;
			Out("D=A");
			Out("@R13");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("A=M");
			Out("M=D");
		}

	//	if(instruction[3] == 'p' && instruction[4] == 'o' && instruction[5] == 'i'){	//pointer
	//		for(int i = 0; i < (strlen(instruction) - 10); i++)
		//		temp[i] = instruction[i + 10];
		//	temp[strlen(instruction) - 10] = '\0';
		if(strcmp(secondword, "pointer") == 0){
			std::ofstream fout(OutputName, std::ios::app);
			fout << "@" << thirdword <<std::endl;
			Out("D=A");
			Out("@3");
			Out("D=D+A");
			Out("@R13");
			Out("M=D");
			Out("@SP");
			Out("M=M-1");
			Out("A=M");
			Out("D=M");
			Out("@R13");
			Out("A=M");
			Out("M=D");

		}

	}

}

void DealWithProgramFlow(){
	switch (CommandType())
	{
	case 112 : {
		AddVMLabel(VMLabelNumber);
		break;
			   }
	case 113 : {
		
		GotoVMLabel(secondword);
		break;
			   }
	case 114 : {
		If_GotoLabel(secondword);
		break;
			   }
	default:
		break;
	}
}

void DealWithFunction(){		//设定开始地址

	strcpy_s(FunctionName, secondword);

	std::ofstream fout(OutputName, std::ios::app);

	fout << "(" << secondword << ")" << std::endl;

	int n = atoi(thirdword);
	for(int i = 0; i < n; i++){
		Out("@0");
		Out("D=A");
		Out("@SP");
		Out("A=M");
		Out("M=D");
		Out("@SP");
		Out("M=M+1");
	}

}

void DealWithCallFunction(){


	std::ofstream fout(OutputName, std::ios::app);
	
	fout << "@" << "return." << ReturnAddressNumber << std::endl;
	Out("D=A");
	Out("@SP");
	Out("A=M");
	Out("M=D");
	Out("@SP");
	Out("M=M+1");

	Out("@LCL");
	Out("D=M");
	Out("@SP");
	Out("A=M");
	Out("M=D");
	Out("@SP");
	Out("M=M+1");

	Out("@ARG");
	Out("D=M");
	Out("@SP");
	Out("A=M");
	Out("M=D");
	Out("@SP");
	Out("M=M+1");

	Out("@THIS");
	Out("D=M");
	Out("@SP");
	Out("A=M");
	Out("M=D");
	Out("@SP");
	Out("M=M+1");

	Out("@THAT");
	Out("D=M");
	Out("@SP");
	Out("A=M");
	Out("M=D");
	Out("@SP");
	Out("M=M+1");

	Out("@SP");
	Out("D=M");
	Out("@5");
	Out("D=D-A");
	int n = atoi(thirdword);
	fout << "@" << n <<std::endl;
	Out("D=D-A");
	Out("@ARG");
	Out("M=D");
	

	Out("@SP");
	Out("D=M");
	Out("@LCL");
	Out("M=D");

	fout << "@" << secondword << std::endl;
	Out("0;JMP");

	fout << "(" << "return." << ReturnAddressNumber << ")" << std::endl;
	ReturnAddressNumber++;
}

void DealWithReturn(){

	Out("@LCL");
	Out("D=M");
	Out("@FRAME");
	Out("M=D");

	Out("@5");
	Out("A=D-A");
	Out("D=M");
	Out("@RET");
	Out("M=D");

	Out("@SP");
	Out("M=M-1");
	Out("A=M");
	Out("D=M");
	Out("@ARG");
	Out("A=M");
	Out("M=D");
	



	Out("@ARG");
	Out("D=M");
	Out("@SP");
	Out("M=D+1");

	Out("@FRAME");
	Out("D=M");
	Out("D=D-1");
	Out("@FRAME");
	Out("M=D");
	Out("A=D");
	Out("D=M");
	Out("@THAT");
	Out("M=D");

	Out("@FRAME");
	Out("D=M");
	Out("D=D-1");
	Out("@FRAME");
	Out("M=D");
	Out("A=D");
	Out("D=M");
	Out("@THIS");
	Out("M=D");

	Out("@FRAME");
	Out("D=M");
	Out("D=D-1");
	Out("@FRAME");
	Out("M=D");
	Out("A=D");
	Out("D=M");
	Out("@ARG");
	Out("M=D");

	Out("@FRAME");
	Out("D=M");
	Out("D=D-1");
	Out("@FRAME");
	Out("M=D");
	Out("A=D");
	Out("D=M");
	Out("@LCL");
	Out("M=D");

	Out("@RET");
	Out("A=M");
	Out("0;JMP");
	
	
}



void AtNormalSymbol(int i){
	std::ofstream fout(OutputName,std::ios::app);
	fout << "@" << SymbolArea << ".";
	fout << i;
	fout << std::endl;
	At_NormalSymbol++;
}

void AddNormalSymbol(int i){
	std::ofstream fout(OutputName,std::ios::app);
	fout << "(" << SymbolArea << "."; 
	fout << i;
	fout << ")";
	fout << std::endl;
	Add_NormalSymbol++;
}

void AddVMLabel(int n){
	
	std::ofstream fout(OutputName, std::ios::app);
	fout << "(" << FunctionName << "$" << secondword << ")" << std::endl;
	Label[VMLabelNumber].LabelNumber = VMLabelNumber;
	strcpy_s(Label[VMLabelNumber].LabelName, secondword);
	VMLabelNumber++;
}

void GotoVMLabel(char* s){
	
	std::ofstream fout(OutputName, std::ios::app);
	fout << "@" << FunctionName << "$" << s << std::endl;
	Out("0;JMP");
}

void If_GotoLabel(char* s){
	
	std::ofstream fout(OutputName, std::ios::app);
	Out("@SP");
	Out("M=M-1");
	Out("A=M");
	Out("D=M");
	fout << "@" << FunctionName << "$" << s << std::endl;
	Out("D;JNE");
}